### PR TITLE
Developed web api xml file generation, web api xml declaration generation, improved action development experience, added PhpPsiElementsUtil

### PR DESCRIPTION
**Pull request by bohdan-harniuk**
**Description** (*)
The main purpose of this PR is to generate web api xml declaration for a service. Generation action is accessible from the context menu if you click on the public method name. Also, it is required for the class to have @api declaration in the annotation. Also, added test for testing this generation.

Screenshot 2021-04-15 at 10 38 15
Screenshot 2021-04-15 at 10 38 33
Screenshot 2021-04-15 at 10 39 25
Screenshot 2021-04-15 at 10 39 48

**Contribution checklist** (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with integration/functional tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        if (method.getContainingClass() == null) {
+            return;
+        }
+
+        final PhpDocComment classDocComment = method.getContainingClass().getDocComment();
+
+        if (!method.getAccess().isPublic()
+                || method.getName().equals("__construct")
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java:51]: We have the same string literal here: `com.magento.idea.magento2plugin.magento.files.Plugin#CONSTRUCT_METHOD_NAME`
Could you please put them in some unified place?

File: src/com/magento/idea/magento2plugin/actions/generation/data/xml/WebApiXmlRouteData.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    private final String moduleName;
+    private final String url;
+    private final String httpMethod;
+    private final String serviceClass;
+    private final String serviceMethod;
+    private final String aclResource;
+
+    /**
+     * Web Api xml declaration DTO constructor.
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/data/xml/WebApiXmlRouteData.java:20]:
```suggestion
     * Web API XML declaration DTO constructor.
```

File: src/com/magento/idea/magento2plugin/actions/generation/dialog/NewWebApiDeclarationDialog.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+     * Create custom components and fill their entries.
+     */
+    @SuppressWarnings({"PMD.UnusedPrivateMethod", "PMD.AvoidInstantiatingObjectsInLoops"})
+    private void createUIComponents() {
+        httpMethod = new ComboBox<>();
+
+        for (final String method : HttpMethod.getHttpMethodList()) {
+            httpMethod.addItem(new ComboBoxItemData(method, method));
+        }
+        aclResource = new FilteredComboBox(new LinkedList<>(Arrays.asList("self", "anonymous")));
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/dialog/NewWebApiDeclarationDialog.java:176]: "self", "anonymous" should be moved to an enum

File: src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import com.magento.idea.magento2plugin.MagentoIcons;
+import com.magento.idea.magento2plugin.actions.generation.dialog.NewWebApiDeclarationDialog;
+import com.magento.idea.magento2plugin.project.Settings;
+import com.magento.idea.magento2plugin.util.php.PhpPsiElementsUtil;
+import org.jetbrains.annotations.NotNull;
+
+public class NewWebApiDeclarationAction extends AnAction {
+
+    public static final String ACTION_NAME = "Create a new WebApi declaration for this method";
+    public static final String ACTION_DESCRIPTION = "Create a new Magento 2 WebApi XML declaration";
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java:24]:
```suggestion
    public static final String ACTION_DESCRIPTION = "Create a new Magento 2 Web API XML declaration";
```

File: src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import com.jetbrains.php.lang.psi.elements.PhpClass;
+import com.magento.idea.magento2plugin.MagentoIcons;
+import com.magento.idea.magento2plugin.actions.generation.dialog.NewWebApiDeclarationDialog;
+import com.magento.idea.magento2plugin.project.Settings;
+import com.magento.idea.magento2plugin.util.php.PhpPsiElementsUtil;
+import org.jetbrains.annotations.NotNull;
+
+public class NewWebApiDeclarationAction extends AnAction {
+
+    public static final String ACTION_NAME = "Create a new WebApi declaration for this method";
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java:23]:
```suggestion
    public static final String ACTION_NAME = "Create a new Web API declaration for this method";
```

File: src/com/magento/idea/magento2plugin/actions/generation/generator/xml/WebApiDeclarationGenerator.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        directoryGenerator = DirectoryGenerator.getInstance();
+        fileFromTemplateGenerator = new FileFromTemplateGenerator(project);
+        getCodeTemplateUtil = new GetCodeTemplateUtil(project);
+        xmlFilePositionUtil = new XmlFilePositionUtil();
+        commonBundle = new CommonBundle();
+        validatorBundle = new ValidatorBundle();
+    }
+
+    /**
+     * Generate WEB API xml declaration.
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/generator/xml/WebApiDeclarationGenerator.java:68]:
```suggestion
     * Generate Web API XML declaration.
```

**Review by VitaliyBoyko — CHANGES_REQUESTED** [review]

File: src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java (hunk 1 above)
Comment (bohdan-harniuk) [reply src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java:51]: Done.

File: src/com/magento/idea/magento2plugin/actions/generation/data/xml/WebApiXmlRouteData.java (hunk 2 above)
Comment (bohdan-harniuk) [reply src/com/magento/idea/magento2plugin/actions/generation/data/xml/WebApiXmlRouteData.java:20]: Done

File: src/com/magento/idea/magento2plugin/actions/generation/dialog/NewWebApiDeclarationDialog.java (hunk 3 above)
Comment (bohdan-harniuk) [reply src/com/magento/idea/magento2plugin/actions/generation/dialog/NewWebApiDeclarationDialog.java:176]: Done.

File: src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java (hunk 4 above)
Comment (bohdan-harniuk) [reply src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java:24]: Done.

File: src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java (hunk 5 above)
Comment (bohdan-harniuk) [reply src/com/magento/idea/magento2plugin/actions/generation/NewWebApiDeclarationAction.java:23]: Done.

File: src/com/magento/idea/magento2plugin/actions/generation/generator/xml/WebApiDeclarationGenerator.java (hunk 6 above)
Comment (bohdan-harniuk) [reply src/com/magento/idea/magento2plugin/actions/generation/generator/xml/WebApiDeclarationGenerator.java:68]: Done.

**Comment by bohdan-harniuk** [comment]
@VitaliyBoyko Requested changes implemented.

Thank you!

**Review by eduard13 — COMMENTED** [review]
Hi @bohdan-harniuk, really good feature.
Additionally, I'm thinking if it worth having some more resource than `anonymous` an `self`. Since we create our own resources via EntityManager, maybe it also makes sense to list all the available `resources`?
cc. @VitaliyBoyko

**Comment by VitaliyBoyko** [comment]
Hi @eduard13 

I thought about it either. 
Let's keep it as it is for the MVP and then extend it with all ACLs. To not overcomplicate this PR.

**Comment by bohdan-harniuk** [comment]
Hello, @VitaliyBoyko ! Thank you for your good suggestion about a useless mandatory requirement to have @api in the class doc. I removed it. Now it looks better.
Thank you!

**Review by VitaliyBoyko — APPROVED** [review]

**Pull request merged**